### PR TITLE
[Feature] MaxCompute(ODPS) Catalog Planner and Optimizer part

### PR DESCRIPTION
**Pull request by dingxin-tech**
Why I'm doing:
The integration between StarRocks and MaxCompute (originally known as ODPS) is crucial for users who need to access their data stored in MaxCompute without the hassle of data migration or duplication. To enhance our users' experience and provide more flexibility in data management, I'm introducing the ODPS Catalog feature to StarRocks.

What I'm doing:
I'm adding a new feature to StarRocks that allows users to connect to ODPS data sources directly. This feature, named ODPS Catalog, acts as a bridge between StarRocks and MaxCompute, giving users the ability to perform queries on MaxCompute datasets from within the StarRocks environment. This integration not only simplifies the data workflow but also ensures that users can leverage the full power of StarRocks's analytics capabilities on their existing data in MaxCompute.

Fixes #issue
#32719

## What type of PR is this:

- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [x] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [x] This pr needs user documentation (for new or modified features or behaviors)
- [x] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:

- [ ] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/planner/OdpsScanNode.java [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+                    break;
+                default:
+                    throw new StarRocksConnectorException(
+                            "unsupported split policy: " + splitsInfo.getSplitPolicy().name());
+            }
+            hdfsScanRange.setOdps_split_info(splitInfo);
+            hdfsScanRange.setUse_odps_jni_reader(true);
+            hdfsScanRange.setFile_length(1);
+            // backend selector use length to load balancing
+            hdfsScanRange.setLength(1);
```
Comment (stephen-shelby) [inline fe/fe-core/src/main/java/com/starrocks/planner/OdpsScanNode.java:129]: Do we have plans to use native to support reading odps data?

File: fe/fe-core/src/main/java/com/starrocks/planner/OdpsScanNode.java [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+    @Override
+    protected String debugString() {
+        MoreObjects.ToStringHelper helper = MoreObjects.toStringHelper(this);
+        helper.addValue(super.debugString());
+        helper.addValue("odpsTable=" + table.getName());
+        return helper.toString();
+    }
+
+    @Override
+    public void finalizeStats(Analyzer analyzer) throws UserException {
```
Comment (stephen-shelby) [inline fe/fe-core/src/main/java/com/starrocks/planner/OdpsScanNode.java:156]: deprecated interface. remove it.

File: fe/fe-core/src/main/java/com/starrocks/planner/OdpsScanNode.java [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+        msg.hdfs_scan_node = tHdfsScanNode;
+    }
+
+    @Override
+    public int getNumInstances() {
+        return scanRangeLocationsList.size();
+    }
+
+    @Override
+    public void computeStats(Analyzer analyzer) {
```
Comment (stephen-shelby) [inline fe/fe-core/src/main/java/com/starrocks/planner/OdpsScanNode.java:201]: ditto

File: gensrc/thrift/PlanNodes.thrift [hunk 4]
```diff
@@ -356,6 +356,10 @@ struct THdfsScanRange {
 
     // identity partition column slots
     18: optional list<Types.TSlotId> identity_partition_slot_ids;
+
+    19: optional bool use_odps_jni_reader
+
+    20: optional map<string, string> odps_split_info
```
Comment (stephen-shelby) [inline gensrc/thrift/PlanNodes.thrift:362]: odps_split_infos?

File: fe/fe-core/src/main/java/com/starrocks/connector/RemoteFileDesc.java [hunk 5]
```diff
@@ -37,10 +37,12 @@ public class RemoteFileDesc {
     // to reduce the memory usage of RemoteFileInfo
     private List<FileScanTask> icebergScanTasks = new ArrayList<>();
     private PaimonSplitsInfo paimonSplitsInfo;
+    private OdpsSplitsInfo odpsSplitsInfo;
 
     private RemoteFileDesc(String fileName, String compression, long length, long modificationTime,
```
[the remark this answers is not in the document]
Comment (zdx19981006) [reply fe/fe-core/src/main/java/com/starrocks/connector/RemoteFileDesc.java:42]: Indeed, perhaps using an Object parameter would be better.

File: fe/fe-core/src/main/java/com/starrocks/connector/odps/OdpsSplitsInfo.java [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+import java.io.ObjectOutputStream;
+import java.io.Serializable;
+import java.util.Base64;
+import java.util.List;
+import java.util.Map;
+
+public class OdpsSplitsInfo {
+    public enum SplitPolicy {
+        /**
+         * 按字节数切分
```
[the remark this answers is not in the document]
Comment (zdx19981006) [reply fe/fe-core/src/main/java/com/starrocks/connector/odps/OdpsSplitsInfo.java:32]: Sure.

File: fe/fe-core/src/main/java/com/starrocks/planner/OdpsScanNode.java (hunk 2 above)
Comment (zdx19981006) [reply fe/fe-core/src/main/java/com/starrocks/planner/OdpsScanNode.java:156]: Done :)

File: fe/fe-core/src/main/java/com/starrocks/connector/odps/OdpsMetadata.java [hunk 7]
```diff
@@ -0,0 +1,416 @@
...
+    }
+
+    public List<RemoteFileInfo> getRemoteFileInfos(Table table, List<PartitionKey> partitionKeys,
+                                                   long snapshotId, ScalarOperator predicate,
+                                                   List<String> columnNames, long limit,
+                                                   TableReadSessionBuilder scanBuilder) {
+        RemoteFileInfo remoteFileInfo = new RemoteFileInfo();
+        OdpsTable odpsTable = (OdpsTable) table;
+        Set<String> set = new HashSet<>(columnNames);
+        List<String> orderedColumnNames = new ArrayList<>();
```
[the remark this answers is not in the document]
Comment (zdx19981006) [reply fe/fe-core/src/main/java/com/starrocks/connector/odps/OdpsMetadata.java:324]: When obtaining splits in ODPS, it is required that the columns passed in must be ordered according to the schema.

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/logical/LogicalOdpsScanOperator.java [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+import com.starrocks.sql.optimizer.operator.OperatorVisitor;
+import com.starrocks.sql.optimizer.operator.ScanOperatorPredicates;
+import com.starrocks.sql.optimizer.operator.scalar.ColumnRefOperator;
+import com.starrocks.sql.optimizer.operator.scalar.ScalarOperator;
+
+import java.util.Map;
+
+public class LogicalOdpsScanOperator extends LogicalScanOperator {
+    private ScanOperatorPredicates predicates = new ScanOperatorPredicates();
+    private boolean hasUnknownColumn;
```
[the remark this answers is not in the document]
Comment (zdx19981006) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/logical/LogicalOdpsScanOperator.java]: now it has been deleted.

File: fe/fe-core/src/main/java/com/starrocks/planner/OdpsScanNode.java (hunk 1 above)
Comment (zdx19981006) [reply fe/fe-core/src/main/java/com/starrocks/planner/OdpsScanNode.java:127]: ODPS has standardized the interface to the complex storage layer using the Arrow format, making the Catalog reading logic almost native. Future plans may involve replacing JNI with a C++-based reading solution and adding the capability of index pushdown. There is currently no plan for direct filesystem reads.

**Comment by zdx19981006** [comment]
Hello,

Some of the commits in this pull request were not signed, causing them to fail the DCO check. I have gone ahead and resubmitted a new pull request with all the necessary signed commits.

You can find the new PR at the following link:

https://github.com/StarRocks/starrocks/pull/37422

Thank you for addressing this issue.

Best regards

**Review by stephen-shelby — DISMISSED** [review]

**Review by stephen-shelby — APPROVED** [review]

**Review by Youngwb — APPROVED** [review]

File: fe/fe-core/src/main/java/com/starrocks/sql/plan/PlanFragmentBuilder.java [hunk 9]
```diff
@@ -1174,6 +1178,57 @@ public PlanFragment visitPhysicalPaimonScan(OptExpression optExpression, ExecPla
             return fragment;
         }
 
+        @Override
+        public PlanFragment visitPhysicalOdpsScan(OptExpression optExpression, ExecPlan context) {
+            LOG.info("[debug] start visit physical odps scan");
```
Comment (packy92) [inline fe/fe-core/src/main/java/com/starrocks/sql/plan/PlanFragmentBuilder.java:1183]: please remove this debug log.

File: fe/fe-core/src/main/java/com/starrocks/sql/plan/PlanFragmentBuilder.java [hunk 10]
```diff
@@ -1174,6 +1178,57 @@ public PlanFragment visitPhysicalPaimonScan(OptExpression optExpression, ExecPla
...
+
+            odpsScanNode.setLimit(node.getLimit());
+
+            tupleDescriptor.computeMemLayout();
+            context.getScanNodes().add(odpsScanNode);
+
+            PlanFragment fragment =
+                    new PlanFragment(context.getNextFragmentId(), odpsScanNode, DataPartition.RANDOM);
+            context.getFragments().add(fragment);
+            LOG.info("[debug] finish visit physical odps scan");
```
Comment (packy92) [inline fe/fe-core/src/main/java/com/starrocks/sql/plan/PlanFragmentBuilder.java:1228]: and this

**Review by packy92 — APPROVED** [review]

File: fe/fe-core/src/main/java/com/starrocks/sql/plan/PlanFragmentBuilder.java (hunk 9 above)
Comment (stephen-shelby) [reply fe/fe-core/src/main/java/com/starrocks/sql/plan/PlanFragmentBuilder.java:1183]: @zdx19981006 will fix it in the next comment

**Pull request merged**